benchmarks/multi-round-qa.py

```diff
@@ -0,0 +1,408 @@
+            self.has_unfinished_request = True
+            return
+
+        if timestamp - self.last_request_time > \
+                self.user_config.gap_between_requests:
+            if self.has_unfinished_request:
+                logger.warning(
+                    f"User {self.user_config.user_id} has an unfinished "
+                    "request and unable to fit the QPS requirement.")
+                return
+
+            logger.debug(f"Issuing the request {self.question_id}")
+            self.last_request_time = timestamp
+            question = self._build_new_question()
+            self.chat_history.on_user_query(question)
+            request_executor.launch_request(self.chat_history,
+                                            self.user_config.answer_len,
+                                            self.on_request_finished)
+            self.has_unfinished_request = True
+            return
+
+    def summary(self) -> pd.DataFrame:
+        df = pd.DataFrame()
+        df["prompt_tokens"] = self.prompt_lengths
+        df["generation_tokens"] = self.generation_lengths
+        df["ttft"] = self.ttfts
+        df["generation_time"] = self.generation_times
+        df["user_id"] = self.user_config.user_id
+        df["question_id"] = range(1, len(self.prompt_lengths) + 1)
+        return df
+
+
+class UserSessionManager:
+
+    def __init__(self, workload_config: WorkloadConfig):
+        self.workload_config = workload_config
+        self.sessions = []
+
+        gap_between_requests_per_user = \
+                workload_config.num_users / workload_config.qps
+        session_alive_time = gap_between_requests_per_user * \
+                (workload_config.num_rounds - 1)
+        self.gap_between_users = session_alive_time / workload_config.num_users
+
+        self.user_id = 0
+        self.last_user_join = 0
+        self.session_summaries = []
+        self.start_time = None
+
+    def _create_user_session(self):
+        self.user_id += 1
+        user_config = UserConfig.new_user_config(self.user_id,
+                                                 self.workload_config)
+        user_session = UserSession(user_config)
+        self.sessions.append(user_session)
+
+    def _remove_finished_sessions(self):
+        sessions_to_remove = [s for s in self.sessions if s.finished]
+        if len(sessions_to_remove) > 0:
+            logger.info(
+                f"Removing {len(sessions_to_remove)} finished sessions, now "
+                f"active users: {len(self.sessions) - len(sessions_to_remove)}"
+            )
+            for session in sessions_to_remove:
+                self.session_summaries.append(session.summary())
+        self.sessions = [s for s in self.sessions if not s.finished]
+
+    def step(self, timestamp: float, executor: RequestExecutor):
+        if self.start_time is None:
+            self.start_time = timestamp
+
+        if timestamp - self.last_user_join > self.gap_between_users:
```

Comment:
Maybe specify that the gap is now fixed and users do not overlap.

Reply:
Users do overlap. It's more like:
```
--------------------------------------------------------> TIME
User 1: req1    req2    req3    req4    req5
User 2:      req1    req2    req3    req4    req5
User 3:            req1    req2    req3    req4    req5
User 4:                  req1    req2    req3    req4    req5
```